src/shell/layer.rs

```diff
@@ -0,0 +1,110 @@
+/// WLR layer abstraction
+use std::cell::Cell;
+use std::rc::Rc;
+
+use wayland_client::{
+    protocol::{wl_output, wl_surface},
+    Attached, Main,
+};
+
+use wayland_protocols::wlr::unstable::layer_shell::v1::client::{
+    zwlr_layer_shell_v1, zwlr_layer_surface_v1,
+};
+
+pub use wayland_protocols::wlr::unstable::layer_shell::v1::client::{
+    zwlr_layer_shell_v1::Layer, zwlr_layer_surface_v1::Anchor,
+};
+
+/// Render event
+
+#[derive(PartialEq, Copy, Clone)]
+pub enum RenderEvent {
+    /// Surface wants a reconfiguration/configuration
+    Configure {
+        /// The new width of the surface
+        width: u32,
+
+        /// The new height of the surface
+        height: u32,
+    },
+
+    /// Surface has closed and wants to be closed here also.
+    Closed,
+}
+
+/// A struct representing the layer surface (wlr)
+
+pub struct LayerSurface {
+    /// The raw wl_surface
+    pub surface: wl_surface::WlSurface,
+
+    pub(crate) layer_surface: Main<zwlr_layer_surface_v1::ZwlrLayerSurfaceV1>,
+
+    /// On what layer it should be positioned
+    pub layer: Layer,
+
+    /// Where on the screen it is positioned
+    pub anchor: Anchor,
+
+    /// The dimensions of the wlr surface
+    pub dimensions: (u32, u32),
+
+    /// The next render event
+    pub render_event: Rc<Cell<Option<RenderEvent>>>,
```

Comment:
I would mimic what `Window` does here with callbacks.

Using a callback here makes sense and also provides the `DispatchData` the client has provided.

This would mean your LayerSurface::new would also now need to take a `FnMut(Event, DispatchData) + 'static` parameter.

Reply:
Could you elaborate?

Reply:
If you look at the methods to create a window you will see one of the parameters is a `Impl` which has type bounds defined in the where statement.

This FnMut the user passes in is essentially a lambda that is invoked when some event occurs on the window.

You'll see where you call quick_assign on the layer you create that you are given an event, that is where you will invoke the callback.

The DispatchData is a dynamic data container for some arbitrary data a user of the library passes in that is available when events come in.